Dataset page update

joshc0044 opened the pull request.

weixuanfu commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx:
```diff
@@ -0,0 +1,199 @@
+    // window.console.debug('median: ', median);
+    // window.console.debug('min (whisker): ', min);
+    // window.console.debug('max (whisker): ', max);
+    // use minimum/max values in data as lower & upper bounds for whisker lines
```
For whisker plot, maybe just simply use minData/maxData as lower/upper bounds. `    let min = q1 - (1.5 * interQuantileRange);
    let max = q3 + (1.5 * interQuantileRange);` may be not needed.

joshc0044 replied:
I made some test plots with plotly.js and have that on a separate branch - https://github.com/EpistasisLab/pennai/tree/plotly_test_for_dataset_page

I posted some screenshots on the slack but just to be thorough I'll upload a few here as well 

![image](https://user-images.githubusercontent.com/6619927/63467215-ec194a00-c432-11e9-87f4-ec1ea3d241d1.png)

The plotly boxplot whisker lines should be configurable; there are three options for specifying a boxplot's boxpoints in plotly - https://plot.ly/javascript/reference/#box-boxpoints

weixuanfu commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx:
```diff
@@ -0,0 +1,199 @@
+    let minData = Math.min(...data_sorted);
+    let maxData = Math.max(...data_sorted);
+
+    // print stats
```
Codes need some clean up before merging.

joshc0044 replied:
I'll remove this and do a bit more cleanup as well

weixuanfu commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx:
```diff
@@ -0,0 +1,199 @@
+    // min = -1;
+    // median = 0;
+    // max = 1;
+    let minData = Math.min(...data_sorted);
```
why not use `minData =  d3.quantile(data_sorted, 0);`? Will the value be not the same?

weixuanfu commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BoxPlot/index.jsx:
```diff
@@ -0,0 +1,199 @@
+      .attr("transform",
+            "translate(" + margin.left + "," + margin.top + ")");
+    // get stats
+    let data_sorted = valByRowObj[rawKey].sort(d3.ascending);
```
To my knowledge, d3.quantile depends on sorted data. Please make sure the data is properly sorted here.

joshc0044 replied:
I believe that is the case as well. 

`valByRowObj` is a javascript object; it's the dataset with the column as a key and the value is an array with all of that columns values. The data is sorted using javascripts native array sort method with d3.ascending - https://github.com/d3/d3-array#ascending

joshc0044 commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarChart/index.jsx:
```diff
@@ -0,0 +1,172 @@
+        if(i < 12) {
+          colorString = d3.schemePaired[i];
+        } else {
+          // if more than 12 classes in dataset use sequential color range
```
also note that if there are more than 12 unique classes then every unique class value color will be from a spectrum - I did this as a way to guarantee each class gets a distinct color (using a spectrum there are infinite colors but the contrast between entries can be small). 

I can also use distinct colors in a range like for the first twelve too

hjwilli commented:
I just rebuilt (`docker-compose build`) from this branch and hard-refreshed the browser, but I'm not seeing the charts.  Maybe you forgot to include a dependency?

![image](https://user-images.githubusercontent.com/1174948/63365859-75068780-c346-11e9-8cd3-54b70b2a5b91.png)

hjwilli commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarCharts/index.jsx:
```diff
@@ -0,0 +1,257 @@
+  render() {
+
+    let plotlyChart = this.createPlotlyStackedChart();
+    plotlyChart.chartCutoff ? plotlyChart = (<p>Too many value to plot</p>) : null;
```
small typo, should be "Too many values to plot"

hjwilli commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarCharts/barChart.test.js:
```diff
@@ -0,0 +1,42 @@
+import BarCharts from './';
```
May want to rename the "BarCharts" component to something along the lines of "StackedBarChart" for clarity

hjwilli commented on lab/webapp/src/components/Dataset/components/DatasetMenu/components/Charts/BarChart/index.jsx:
```diff
@@ -0,0 +1,203 @@
+    // for every entry in depColSet, map keys to color
+    let colorObj = {};
+    let colorList = [];
+    let colorStringList = [
```
Identical values for the color list in `colorStringList` and the color-to-name mapping variable are defined in both `BarChart/index.jsx` and `BarCharts/index.jsx`.  This color scheme should probably be defined once elseware and referenced from these components.

It's likely we will want to reference this color scheme for plots on other pages later.

joshc0044 commented on lab/webapp/package.json:
```diff
@@ -4,7 +4,7 @@
   "scripts": {
     "build": "webpack",
-    "build-dev": "webpack --watch",
+    "build-dev": "webpack --watch --mode=development",
```
I believe the only way to invoke/call build-dev is connecting directly into the running instance of pennai_lab_1 through `docker exec` and and then manually running the build-dev command; in a production environment this should normally not occur but I can remove/revert this change if deemed prudent

hjwilli replied:
You are correct, this line is fine.

joshc0044 commented:
I fixed the typo for "too many values", renamed `BarCharts` to `StackedBarChart` and moved the color list & obj to a seperate static file. 

There was also a mistake for selecting alternative colors when were more than 12 classes - a fix for this is also included in this commit, I tested it with the kykopt dataset from pmlb dataset with 18 distinct classes and each bar has its own color

joshc0044 commented:
I fixed the problems with the boxplots & stacked bar charts

For point 2 in Weixuan's comment above, the max number of values the stacked bar chart can display is currently set to 40. This limit is not technically necessary as the chart should theoretically work with an arbitrary number of unique values for that given category. But, for the sake of readability, there is an upper limit. If an alternative chart is preferable to the message 'Too many values to plot' that should be feasible as well; create a pie chart or something else deemed appropriate. 

And for point 3 - apparently it is currently not possible to modify the styling of the content displayed when hovering over the box plot (https://community.plot.ly/t/box-plot-and-hover-info/10341/2). The content of the hover text is different depending on whether the boxplot is just showing the mean or the mean and the standard deviation (https://plot.ly/javascript/box-plots/#box-plot-styling-mean-and-standard-deviation)

lacava commented:
- [ ] styling comments:
    - [ ] the text is too small and there is too much whitespace

- [ ] boxplots
    - [ ] the SD diamons are *kind of* cool, but probably just confusing since they aren't used regularly. I would just show the mean dotted line. 
    - [ ] the xticks are too small
    - [ ] i think it would look better with a border on the box plots
    - [ ] aspect ratio: the plots should be shorter and wider so that you can see more plots in browser at a time, much like openML
    - [ ] remove extraneous buttons: toggle spike (what does this do), show closest data on hover, compare data on hover

hjwilli commented:
- [ ] display the number of unique values in or next to the 'type' column
- [ ] for numeric fields, if the number of unique values is small (<= 5), use stacked bar charts instead of box plots

joshc0044 commented:
I'm not 100% sure I understand how to take a stacked bar chart and display a pie chart. In the following example I'm using the `diabetes_small.csv` which I have posted on slack on aug 22 and will post here as well.

![image](https://user-images.githubusercontent.com/6619927/65534571-c1a52b80-decd-11e9-9508-dd1e823bacc9.png)

For this stacked chart there are 16 unique values in the dataset column for `payer_code`. If I understand the request correctly, the five most numerous values will be slices in the pie chart: `MC, ?, HM, BC, SP` and the rest will be combined into a single slice. Is it correct to take the 'stacks' of the bar chart, consolidate/combine them into one total, and display them as pie chart slices? And I will not incorporate the dependent column/target class info.

If this is not the case please clarify with this example what info from the stacked chart should be displayed in the pie chart

joshc0044 commented:
here is the small test dataset I used -
[diabetes_small.zip](https://github.com/EpistasisLab/pennai/files/3648469/diabetes_small.zip)
with the following dependent/target class: `readmitted`
with the following categorical features:
```
race,gender,age,weight,payer_code,medical_specialty,diag_1,diag_2,diag_3,max_glu_serum,A1Cresult,metformin,repaglinide,nateglinide,chlorpropamide,glimepiride,acetohexamide,glipizide,glyburide,tolbutamide,pioglitazone,rosiglitazone,acarbose,miglitol,troglitazone,tolazamide,examide,citoglipton,insulin,glyburide.metformin,glipizide.metformin,glimepiride.pioglitazone,metformin.rosiglitazone,metformin.pioglitazone,change,diabetesMed
```

joshc0044 commented:
as a quick first pass/rough draft here is a pie chart which should be similar to how it will look -
![image](https://user-images.githubusercontent.com/6619927/65540173-db983b80-ded8-11e9-859a-5383c0321d1e.png)

joshc0044 commented:
for the chart button menu 
>   * [ ]  remove extraneous buttons: toggle spike (what does this do), show closest data on hover, compare data on hover

the `compare data on hover` button looks like this for bar charts -
![image](https://user-images.githubusercontent.com/6619927/65707471-9c91f380-e05a-11e9-808a-b02c1ea14a40.png)
the `closest data on hover button` looks like this - 
![image](https://user-images.githubusercontent.com/6619927/65707530-bb908580-e05a-11e9-93dc-b139f424c8aa.png)

for boxplots `compare data on hover` it looks like -
![image](https://user-images.githubusercontent.com/6619927/65708035-ac5e0780-e05b-11e9-92aa-d9acf19acbb0.png)

and for `closest data on hover button` boxplots - 
![image](https://user-images.githubusercontent.com/6619927/65708100-d283a780-e05b-11e9-8c81-d0a1e48892a0.png)

and `toggle spike lines` does this -
![image](https://user-images.githubusercontent.com/6619927/65708796-61dd8a80-e05d-11e9-808f-d7c2495f7ea5.png)

I'll disable the spike line button and the compare data buttons for the box plots; these buttons for this chart do not do much. For the stacked bar charts though, I can remove them or leave them as they have more actual utility

hjwilli commented:
Archiving this pull request.
